compiler/pipes/check-type-hint-variance.cpp

```diff
@@ -0,0 +1,82 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "compiler/pipes/check-type-hint-variance.h"
+#include "compiler/data/class-data.h"
+#include "compiler/data/function-data.h"
+#include "compiler/type-hint.h"
+
+#include "common/containers/final_action.h"
+
+namespace {
+
+FunctionPtr get_derived_method(FunctionPtr f, ClassPtr derived_class) {
+  if (f->modifiers.is_static()) {
+    auto *member = derived_class->members.get_static_method(f->local_name());
+    return member ? member->function : FunctionPtr{};
+  }
+  auto *member = derived_class->members.get_instance_method(f->local_name());
+  return member ? member->function : FunctionPtr{};
+}
+
+} // namespace
+
+void CheckTypeHintVariance::execute(FunctionPtr f, DataStream<FunctionPtr> &os) {
```

Comment:
It's really hard to understand, why some checks are made here, in a separate pipe, and some in `check_signatures_variance()` — in a function named the same, but in another file and another pipe... Can't they be joined? Can you please add a comment of such dictinction?

Reply:
There is a comment above the pipe class.

![image](https://user-images.githubusercontent.com/6286655/116259098-2e556b80-a77e-11eb-8b2e-7b3c0b0c53e3.png)


The reason is: we merge phpdoc and typehints after the parse-and-apply-phpdocs.
To avoid missing some incompatible signatures that are only compatible due to the phpdocs (which PHP ignores) we have an extra pass that verifies that we follow the PHP rules and will require type hints if they're needed.
So that pipe checks "whether you should add a type hint here" and "whether you can't use a type hint here". It doesn't really check types compatiblity.

Later, in virtual-method-generator.cpp we use the combined type info to check if signatures are actually compatible. So we're probably more strict than PHP, but that the intent: we don't want the code to compile as KPHP and fail as PHP, but we don't have a goal of 100% strict PHP-to-KPHP match here. It's OK to reject some badly typed code for us.